tools/autograd/gen_variable_type.py

```diff
@@ -782,6 +806,42 @@ def emit_check_no_requires_grad(tensor_args, args_with_derivatives):
             body.append('check_no_requires_grad({}, "{}");'.format(name, name))
         return body
 
+    def emit_native_body(declaration):
```

Comment:
I'm scared about the amount of code duplication going on between emit_native_body and emit_body. It will be very easy for changes in one pathway to not be picked up correctly in the other.

Reply:
I guess the primary objection here is the "duplication" in `return_value` and `trace_outputs` computation? In the case of list return, they are not really duplicates at all in their current form, and this has to do with how I decided to do the codegen. Compare:

```
// Native dispatch
std::vector<Tensor> VariableType::chunk(const Tensor & self, int64_t chunks, int64_t dim) const {
    profiler::RecordFunction profiler("chunk");
    auto ret = Type::chunk(self, chunks, dim);
    if (jit::tracer::isTracing( self )) {
      jit::Node *n = jit::tracer::recordTrace( "chunk", { self }, cast_tensor_list(ret) );
      setattr(n, jit::stringToSymbol("chunks"), chunks);
      setattr(n, jit::stringToSymbol("dim"), dim);
    }
    return ret;
}

// Non-native function
std::vector<Tensor> VariableType::split(const Tensor & self, int64_t split_size, int64_t dim) const {
    profiler::RecordFunction profiler("split");
    auto& self_ = unpack(self, "self", 0);
    std::shared_ptr<SplitBackward> grad_fn;
    auto flags = compute_flags({ self });
    if (flags.requires_grad) {
      grad_fn = std::make_shared<SplitBackward>();
      grad_fn->next_functions = compute_next_functions({ self });
      grad_fn->self_sizes = self.sizes();
      grad_fn->self_ = SavedVariable(self, false);
      grad_fn->split_size = split_size;
      grad_fn->dim = dim;
    }
    auto ret = as_variable(baseType->split(self_, split_size, dim));
    set_flags(ret, flags, grad_fn);
    if (jit::tracer::isTracing( self )) {
      jit::Node *n = jit::tracer::recordTrace( "split", { self }, ret );
      setattr(n, jit::stringToSymbol("split_size"), split_size);
      setattr(n, jit::stringToSymbol("dim"), dim);
    }
    return as_tensor_list(ret);
}
```

The reason why you can't just reuse the computation here is because the types of `ret` codgenned here are different. `ret` in the native case is a `std::vector<Tensor>`, but it is a `std:vector<Variable>` in the non-native case (due to the `as_variable` computation.)

I suppose we could merge the codepaths by renaming `ret` and then adding an impedence matching assignment:

```
// Native dispatch
std::vector<Tensor> VariableType::chunk(const Tensor & self, int64_t chunks, int64_t dim) const {
    profiler::RecordFunction profiler("chunk");
    auto var_ret = Type::chunk(self, chunks, dim); // var_ret is std::vector<Tensor>
    if (jit::tracer::isTracing( self )) {
      auto ret = cast_tensor_list(var_ret); // recordTrace requires ArrayRef<Variable>
      jit::Node *n = jit::tracer::recordTrace( "chunk", { self }, ret );
      setattr(n, jit::stringToSymbol("chunks"), chunks);
      setattr(n, jit::stringToSymbol("dim"), dim);
    }
    return var_ret;
}
```

And then, to use the same codepath, I suppose the native template could ignore all of the extra bits (variable flags setting, variable wrapping, etc.) which it doesn't need? Somehow this doesn't seem more clear to me.